File: moo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
   _____________________
  /   _ __   ___  ___   \\
  |  | '  \ / _ \/ _ \  |
  \  |_|_|_|\___/\___/  /
   ---------------------
          \   ^__^
           \  (oo)\_______
              (__)\       )\/\\
                  ||----w |
                  ||     ||
  Editor-agnostic markdown live preview server.

Usage: moo [options] FILE

Options:
  -q, --quiet           Quiet mode.
  --debug               Enable server debug log.
  -p PORT, --port=PORT  Server port. Random port by default.

"""

# import bottle
# import subprocess
# from multiprocessing import Process
from gevent import monkey; monkey.patch_all()
import sys
import os
import misaka
import logging
import time
import pygments
import socket
from cgi import escape as escape_html
from docopt import docopt
from bottle import get, delete, view, response, static_file, route, run
from pygments.lexers import get_lexer_by_name
from pygments.formatters import HtmlFormatter
# if chardet is available, use it to detect file encoding
# try:
#     import chardet1 as chardet
#     # FIXME chardet returns None!
#     charset_detect = lambda x: chardet.detect(x)['encoding']
# except ImportError:
#     logging.warning('chardet not found, use UTF-8 as file encoding instead')
# UTF-8 as fallback encoding
charset_detect = lambda x: 'utf-8'

# inteval to check file updates (unit: second)
RELOAD_INTERVAL = 1
# file to preview
MARKDOWN_FILE = ''
# misaka renderer extensions
RENDER_EXTENSIONS = misaka.EXT_FENCED_CODE | \
                    misaka.EXT_NO_INTRA_EMPHASIS | \
                    misaka.EXT_AUTOLINK | \
                    misaka.EXT_STRIKETHROUGH | \
                    misaka.EXT_LAX_HTML_BLOCKS | \
                    misaka.EXT_SUPERSCRIPT | \
                    misaka.EXT_TABLES
# misaka HTML flags
HTML_FLAGS = misaka.HTML_TOC

class HtmlRenderer(misaka.HtmlRenderer, misaka.SmartyPants):
    def block_code(self, text, lang):
        if not lang:
            return '\n<pre><code>%s</code></pre>\n' % \
                escape_html(text.strip())
        lexer = get_lexer_by_name(lang, stripall=True)
        formatter = HtmlFormatter()
        return pygments.highlight(text, lexer, formatter)

renderer = HtmlRenderer(HTML_FLAGS)
to_html = misaka.Markdown(renderer, RENDER_EXTENSIONS).render

def render_markdown():
    title = os.path.basename(MARKDOWN_FILE)
    try:
        with open(MARKDOWN_FILE) as f:
            raw_content = f.read()
            decoded_content = raw_content.decode(charset_detect(raw_content))
            rendered = to_html(decoded_content)
            return title, rendered
    except IOError, e:
        logging.error(e.strerror)
        raise e


@get('/')
@view('preview')
def index():
    title, content = render_markdown()
    return { 'content': content, 'title': title }


@delete('/')
def index():
    logging.info('server shutdown')
    sys.exit(0)


@route('/static/style.css')
def stylesheet():
    return static_file('style.css', root='static')

def reloader(filename, interval):
    # last modified time
    mtime = None
    while True:
        try:
            new_mtime = os.stat(filename).st_mtime
        except OSError, e:
            logging.critical('%s: %s', MARKDOWN_FILE, e.strerror)
            sys.exit(-1)

        if mtime is None:
            mtime = new_mtime
            continue
        elif new_mtime > mtime:
            mtime = new_mtime
            yield 'data: updated\n\n'
            logging.info('detected file change in %s', MARKDOWN_FILE)
        time.sleep(interval)


@get('/update')
def update():
    logging.debug('sending event source stream...')
    response.content_type = 'text/event-stream'
    return reloader(MARKDOWN_FILE, RELOAD_INTERVAL)

# class MooServer(Process):
#     def __init__(self, *args, **kwargs):
#         super(MooServer, self).__init__()
#         self.args = args
#         self.kwargs = kwargs
#     def run(self):
#         logging.info('starting server at port %d' % PORT)
#         bottle.run(*self.args, **self.kwargs)

def pick_unused_port():
    s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
    s.bind(('127.0.0.1', 0))
    addr, port = s.getsockname()
    s.close()
    return port

def open_local_url(port):
    import webbrowser
    logging.debug('opening web browser...')
    # from ipdb import set_trace; set_trace()
    local_url = 'http://127.0.0.1:%d' % port
    webbrowser.open_new(local_url)
    # if sys.platform == 'win32':
    #     # Windows
    #     subprocess.Popen(['start', local_url], shell=True)
    # elif sys.platform == 'darwin':
    #     # Macintosh
    #     subprocess.Popen(['open', local_url])
    # else:
    #     try:
    #         # Linux with X window system
    #         subprocess.Popen(['xdg-open', local_url])
    #     except OSError:
    #         # Oops
    #         logging.critical('cannot open preview URL')
    #         sys.exit(-1)

if __name__ == '__main__':
    args = docopt(__doc__, version='0.2.0')

    MARKDOWN_FILE = args['FILE']

    # shut-your-mouth-up mode
    if args['--quiet']:
        logging_level = logging.ERROR
    # logging configs
    if args['--debug']:
        logging_level = logging.DEBUG
        server_quiet = False
        server_debug = True
    else:
        logging_level = logging.INFO
        server_quiet = True
        server_debug = False

    logging.basicConfig(level=logging_level, format='%(asctime)s %(levelname)-8s %(message)s',
            datefmt='%Y-%m-%d %H:%M:%S')

    if args['--port'] is None:
        # random port
        PORT = pick_unused_port()
    else:
        try:
            PORT = int(args['--port'])
        except:
            sys.stderr.write('Invalid port number\n')
            sys.exit(-1)

    # open preview URL in browser
    open_local_url(PORT)

    # start server
    run(port=PORT, quiet=server_quiet, debug=server_debug, server='gevent')

File: moo.js
(function() {
  var Moo, colors, commander, config, configure, error, examples, exports, express, fs, hljs, init, jade, marked, open, pandoc, parse, path, preview, run, serve, timeLog, version, warn, watch, _,
    __slice = [].slice,
    __indexOf = [].indexOf || function(item) { for (var i = 0, l = this.length; i < l; i++) { if (i in this && this[i] === item) return i; } return -1; };

  express = require('express');

  path = require('path');

  fs = require('fs');

  jade = require('jade');

  open = require('open');

  colors = require('colors');

  marked = require('marked');

  commander = require('commander');

  _ = require('lodash');

  watch = require('chokidar').watch;

  hljs = require('highlight.js');

  error = function(msg) {
    return console.error(("" + 'error'.red + " - " + msg).bold);
  };

  warn = function(msg) {
    return console.warn(("" + 'warning'.magenta + " - " + msg).bold);
  };

  timeLog = function(msg) {
    return console.log("" + ((new Date).toLocaleTimeString()) + " - " + msg);
  };

  preview = function(options) {
    if (options == null) {
      options = {};
    }
    configure(options);
    if (config.sources.length !== 1) {
      error('expect one input file');
    }
    return serve();
  };

  serve = function() {
    var app, listen, server, source;

    source = config.sources[0];
    server = null;
    app = express();
    app.set('views', config.layout);
    app.set('view engine', 'jade');
    app.use(express.errorHandler());
    app.use('/', express["static"](path.dirname(fs.realpathSync(source))));
    app.use('/__moo__', express["static"](config.resources));
    app.get('/', function(req, res) {
      return res.render('index.jade');
    });
    app.get('/content', function(req, res) {
      return fs.readFile(source, function(err, buffer) {
        var text;

        if (err) {
          error(err.message);
          throw err;
        }
        text = buffer.toString();
        return parse(source, text, function(err, content) {
          if (err) {
            console.error(err);
            throw err;
          }
          return res.json(content);
        });
      });
    });
    app.get('/update-event', function(req, res) {
      var watcher;

      req.socket.setTimeout(0);
      res.socket.setTimeout(0);
      res.set({
        'Content-Type': 'text/event-stream',
        'Cache-Control': 'no-cache',
        'Connection': 'keep-alive'
      });
      watcher = watch(source, {
        persistent: true
      });
      watcher.on('change', function() {
        timeLog("changed: " + source);
        return res.write('data: updated\n\n');
      });
      watcher.on('unlink', function() {
        error("removed: " + source);
        watcher.close();
        return server.close();
      });
      watcher.on('error', function(err) {
        error(err.message);
        throw err;
      });
      return res.on('close', function() {
        return watcher.close();
      });
    });
    listen = function(port) {
      var url;

      server = app.listen(port);
      url = "http://localhost:" + (server.address().port);
      console.log("Server listening at " + url.blue.bold + ". Press Ctrl-C to stop.");
      open(url);
      return server;
    };
    return listen(config.port);
  };

  exports = function(options) {
    var files, nextFile, template, write;

    if (options == null) {
      options = {};
    }
    configure(options);
    template = "" + config.layout + "/export.jade";
    template = jade.compile(fs.readFileSync(template), {
      filename: template
    });
    files = config.sources.slice();
    write = function(source, content) {
      var destination, html;

      destination = function(file) {
        return path.join(path.dirname(file), path.basename(file, path.extname(file)) + '.html');
      };
      html = template(content);
      console.log("moo: " + source + " -> " + (destination(source)));
      return fs.writeFileSync(destination(source), html);
    };
    nextFile = function() {
      var source;

      source = files.shift();
      return fs.readFile(source, function(err, buffer) {
        var text;

        if (err) {
          throw err;
        }
        text = buffer.toString();
        return parse(source, text, function(err, content) {
          if (err) {
            throw err;
          }
          write(source, content);
          if (files.length) {
            return nextFile();
          }
        });
      });
    };
    return nextFile();
  };

  parse = function(source, text, cb) {
    var first, hasTitle, html, stripFrontmatter, title, tokens;

    title = path.basename(source);
    hasTitle = false;
    if (config.pandoc) {
      return pandoc(title, text, cb);
    }
    stripFrontmatter = function() {
      var mainmatter, match;

      if (/^-{3}\n/.test(text)) {
        mainmatter = text.split('---\n');
        mainmatter.shift();
        if (mainmatter.length >= 2) {
          match = /(?:^|\n)title:\s*(.+)/i.exec(mainmatter.shift());
          if (match) {
            hasTitle = true;
            title = match[1];
          }
          return text = mainmatter.join('---\n');
        }
      }
    };
    stripFrontmatter();
    tokens = marked.lexer(text);
    if (!hasTitle) {
      first = tokens[0];
      if (first && first.type === 'heading' && first.depth === 1) {
        title = first.text;
      }
    }
    html = marked.parser(tokens);
    return cb(null, {
      title: title,
      html: html
    });
  };

  pandoc = function(title, text, cb) {
    var options, pdc;

    pdc = require('pdc');
    options = ['--highlight-style', 'pygments'];
    return pdc(text, 'markdown', 'html', options, function(err, result) {
      if (err) {
        return cb(err, null);
      } else {
        return cb(null, {
          title: title,
          html: result
        });
      }
    });
  };

  config = {
    layout: 'github',
    port: 0,
    pandoc: false
  };

  configure = function(options) {
    _.extend(config, _.pick.apply(_, [options].concat(__slice.call(_.keys(config)))));
    config.resources = "" + __dirname + "/resources";
    config.layout = "" + config.resources + "/" + config.layout;
    return config.sources = options.args.sort();
  };

  init = function() {
    var highlight, l, languages;

    languages = (function() {
      var _results;

      _results = [];
      for (l in hljs.LANGUAGES) {
        _results.push(l);
      }
      return _results;
    })();
    highlight = hljs.highlight;
    return marked.setOptions({
      smartypants: true,
      highlight: function(code, lang) {
        if (__indexOf.call(languages, lang) >= 0) {
          return highlight(lang, code).value;
        }
        return code;
      }
    });
  };

  examples = ["  Examples:", "", "  - preview:".yellow.bold, "    + live preview markdown in the web browser", "    + example: " + '$ moo README.md'.cyan.bold, "  - export:".yellow.bold, "    + export markdown files to html docs", "    + example: " + '$ moo -e ch1.md ch2.md ch3.md'.cyan.bold, ""].join('\n');

  version = JSON.parse(fs.readFileSync("" + __dirname + "/package.json")).version;

  run = function(args) {
    var c;

    if (args == null) {
      args = process.argv;
    }
    c = config;
    commander.version(version).usage('[options] files').on('--help', function() {
      return console.log(examples);
    }).option('-e, --export', 'export mode').option('-p, --port <port>', 'server port (default: random)', parseInt, c.port).option('--pandoc', 'use pandoc markdown converter').parse(args).name = "moo";
    if (commander.args.length) {
      init();
      if (commander["export"]) {
        return exports(commander);
      } else {
        return preview(commander);
      }
    } else {
      return console.log(commander.helpInformation());
    }
  };

  Moo = module.exports = {
    run: run,
    preview: preview,
    exports: exports,
    parse: parse
  };

}).call(this);
